src/extractor/intersection/intersection_analysis.cpp

```diff
@@ -0,0 +1,808 @@
+                  const EdgeBasedNodeDataContainer &node_data_container,
+                  const std::vector<util::Coordinate> &node_coordinates,
+                  const extractor::CompressedEdgeContainer &compressed_geometries,
+                  const RestrictionMap &node_restriction_map,
+                  const std::unordered_set<NodeID> &barrier_nodes,
+                  const guidance::TurnLanesIndexedArray &turn_lanes_data,
+                  const IntersectionEdge &incoming_edge)
+{
+    const auto intersection_node = graph.GetTarget(incoming_edge.edge);
+    const auto &outgoing_edges = intersection::getOutgoingEdges(graph, intersection_node);
+    auto edge_geometries = getIntersectionOutgoingGeometries(
+        graph, compressed_geometries, node_coordinates, intersection_node);
+
+    // Add incoming edges with reversed bearings
+    const auto edges_number = edge_geometries.size();
+    edge_geometries.resize(2 * edges_number);
+    for (std::size_t index = 0; index < edges_number; ++index)
+    {
+        const auto &geometry = edge_geometries[index];
+        const auto remote_node = graph.GetTarget(geometry.edge);
+        const auto incoming_edge = graph.FindEdge(remote_node, intersection_node);
+        edge_geometries[edges_number + index] = {incoming_edge,
+                                                 util::bearing::reverse(geometry.initial_bearing),
+                                                 util::bearing::reverse(geometry.perceived_bearing),
+                                                 geometry.length};
+    }
+
+    // Enforce ordering of edges by IDs
+    std::sort(edge_geometries.begin(), edge_geometries.end());
+
+    return convertToIntersectionView(graph,
+                                     node_data_container,
+                                     node_restriction_map,
+                                     barrier_nodes,
+                                     edge_geometries,
+                                     turn_lanes_data,
+                                     incoming_edge,
+                                     outgoing_edges,
+                                     std::unordered_set<EdgeID>());
+}
+
+IntersectionEdge skipDegreeTwoNodes(const util::NodeBasedDynamicGraph &graph, IntersectionEdge road)
+{
+    std::unordered_set<NodeID> visited_nodes;
+    (void)visited_nodes;
+
+    // Skip trivial nodes without generating the intersection in between, stop at the very first
+    // intersection of degree > 2
+    const auto starting_node = road.node;
+    auto next_node = graph.GetTarget(road.edge);
+    while (graph.GetOutDegree(next_node) == 2 && next_node != starting_node)
+    {
+        BOOST_ASSERT(visited_nodes.insert(next_node).second);
+        const auto next_edge = graph.BeginEdges(next_node);
+        road.edge = graph.GetTarget(next_edge) == road.node ? next_edge + 1 : next_edge;
+        road.node = next_node;
+        next_node = graph.GetTarget(road.edge);
+    }
```

Comment:
This check always returns true because of an assignment  `query_edge = next_edge;` one line above and it was a comparison  of two identical data structures.